rfd/0156-notifications.md

```diff
@@ -0,0 +1,431 @@
+      },
+   },
+}
+
+notification := types.Notification{
+   Type: notifications.IntegrationMisconfiguredType,
+   Metadata: types.Metadata{
+      Name: "Integration Misconfigured",
+      Description: fmt.Sprintf("The integration [%s] was misconfigured.", integration.name),
+      Labels:      map[string]string{"integration-id": integration.id},
+   },
+}
+
+notifyUsersByPermissions(roleConditions, notification)
+```
+
+### Flow
+
+#### Web UI
+
+Upon logging in and on every page load, a request will be made to fetch all the user's notifications. This fetch will be performed asynchronously so that the login or page loading process is not slowed down by having to wait for the notifications response.
+The notifications service will query the database and return the list of the user's notifications after filtering out those not meant for them. In the case of global notifications, only notifications with matchers that match the user (ie. by role or
+permissions) will be kept. Notifications that the user dismissed will also be filtered out.
+
+This is a simplified diagram of the flow:
+
+```mermaid
+sequenceDiagram
+   participant WebUI
+   participant ProxyAuth as Proxy & Auth
+   participant BackendDB as Backend DB
+
+   WebUI->>+ProxyAuth: Fetch notifications on login or page load
+   ProxyAuth->>+BackendDB: Query database for notifications and user's notification states
+   BackendDB-->>-ProxyAuth: Return notifications and user's notification states
+   ProxyAuth->>ProxyAuth: Filter out notifications that don't target the user or that were dismissed
+   ProxyAuth-->>-WebUI: Return filtered notifications
+```
+
+###### Stretch goal
+
+In the future, we can implement the use of WebSockets alongside a database watcher in order to allow for real-time notifications, this would work by only performing one initial fetch on login for all notifications and storing them in `localStorage`,
+then, any new notifications received during the user's session will be sent to the user's browser in real-time via a WebSocket connection.
+
+We could also look into implementing the use of a service worker that can run and listen for notifications even if no tabs are open, this could also work with the [Push API](https://developer.mozilla.org/en-US/docs/Web/API/Push_API) and
+[Notifications API](https://developer.mozilla.org/en-US/docs/Web/API/Notifications_API) to send push notifications from the browser. However, this would introduce some potential security-related issues and would require more consideration.
+
+#### tsh
+
+The first version will not include `tsh` support for notifications. This can be explored in the future, and would likely be limited to a `tsh notifications ls` command for listing notifications.
+
+### UX
+
+Most of the details regarding the UX are still being discussed and have yet to be finalized, they are being tracked in [this GitHub issue](https://github.com/gravitational/cloud/issues/6486). The details proposed below are subject to change.
```

Comment:
This is not entirely true. UX is not "how it looks" but "how it works" - and you've alluded to some of that above.

I think it would be good to use this section to discuss how the user will interact with the notification system.

- 1 or more notifications are created
- user logs in and sees a badge with "N" new notifications.
- each notification displays an indicator identifying it as new and previously unseen
- now that all N of these notifications have been viewed, the state is updated in the backend. if the user were to refresh the page, they would no longer show as unread.
- user clicks to interact with one of the notifications. after some custom action, that notification now renders differently (does it?)
- user clicks dismiss on one of the notifications. they never see it again

etc.


EDIT: I see you discussed almost all of this below in the Web UI section. Nice job.